### PR TITLE
add option to only attempt copy command on source url

### PR DESCRIPTION
**Pull request by gabemontero**
Fixes https://github.com/openshift/source-to-image/issues/367

@bparees @mfojtik PTAL

### CODE REVIEW

File: cmd/s2i/main.go [hunk 1]
```diff
@@ -181,6 +181,7 @@ $ s2i build . centos/ruby-22-centos7 hello-world-app
 	buildCmd.Flags().StringSliceVar(&(cfg.DropCapabilities), "cap-drop", []string{}, "Specify a comma-separated list of capabilities to drop when running Docker containers")
 	buildCmd.Flags().StringVarP(&(oldDestination), "location", "l", "",
 		"DEPRECATED: Specify a destination location for untar operation")
+	buildCmd.Flags().BoolVarP(&(cfg.ForceCopy), "copy", "c", false, "Only attempt a local file system copy against the source URL")
```
Comment (bparees) [inline cmd/s2i/main.go]: "Use local file system copy instead of git cloning the source url"

File: docs/cli.md [hunk 2]
```diff
@@ -72,6 +72,7 @@ that image and add them to the tar streamed to the container into `/artifacts`.
 | Name                       | Description                                             |
 |:-------------------------- |:--------------------------------------------------------|
 | `--callback-url`           | URL to be invoked after a successful build (see [Callback URL](#callback-url)) |
+| `-c (--copy)`             | Only attempt a local file system copy against the source URL (allows for inclusion of empty directories) |
```
Comment (bparees) [inline docs/cli.md]: "Use local file system copy instead of git cloning the source URL (allows...."

**Comment by bparees** [comment]
one wording change suggestion and lgtm

**Comment by gabemontero** [comment]
word changes pushed

**Comment by bparees** [comment]
lgtm [merge]

**Comment by rhcarvalho** [comment]
@gabemontero where does the "copy" actually happen?

**Comment by gabemontero** [comment]
@rhcarvalho it happens here:  https://github.com/openshift/source-to-image/blob/master/pkg/scm/file/download.go#L28

**Comment by rhcarvalho** [comment]
Hmm I was surprised that doesn't appear in this patch. Looking at the changes seems that the meat was at https://github.com/openshift/source-to-image/pull/428/files#diff-e9f1cf6224310cfa91dbb395a6db1f85R38 but then I could not understand how this affects the code path. Some comments advertise that `cp` would be used instead of `git clone`, so I'm guessing the two mechanisms were already in place and we're just exposing it here, did I get that right?

**Comment by bparees** [comment]
>  so I'm guessing the two mechanisms were already in place and we're just exposing it here, did I get that right?

right.  we're giving you a way to override the magic that picks the mechanism.

**Comment by rhcarvalho** [comment]
> we're giving you a way to override the magic that picks the mechanism.

I couldn't find where we map `forceCopy = true` to "make sure we use `cp` and not `git clone`"?

The condition reads `details.FileExists && (details.UseCopy || forceCopy)`, so I'm not clear on how `details.UseCopy` gets set or what it means, but I read that as "if the file exists and either UseCopy or forceCopy is true, then return a File instance" -- but no guarantee what will be the mechanism?

Perhaps what happens is that, for empty dirs, `git clone` is not possible so `cp` is the only option left.
But if I use `--force-copy` with a directory containing `.git`, won't it just do a `git clone`?!

**Comment by rhcarvalho** [comment]
The option name is actually `--copy`, my bad.

**Pull request merged**